### DISCUSSION TITLE
Use eager loaded relations to determine whether or not to show conditional properties of a JSON Resource

### DISCUSSION THREAD

**Discussion by jamiethorpe** [dedoc/scramble#151] [Ideas]
Currently, conditional properties of a JSON Resource will either display or not display `Required` if they use methods like `whenLoaded`. I think it would be a great idea to take this a step further and check the controller to see if any relations are eager loaded. If they are, the docs would compare that to the properties using the `whenLoaded` method and only display the ones that match. This would make the docs much easier to read from the perspective of the developer consuming them.

**Comment by romalytvynenko** [comment]
@jamiethorpe can you please share some code examples of how exactly your controllers (I guess) looks like?

I'm interested in the part of how you load relationships, so I can ship this feature!

**Reply by jamiethorpe** [reply]
Absolutely. Here's an example:

```php
public function index(IndexTransactionRequest $request)
{
    try {
        $transactions = Transaction::with(['currentStage', 'ledgerEntries'])
            ->filter(new TransactionFilter($request->all()))
            ->get();

        return TransactionResource::collection($transactions);

    } catch (\Exception $e) {
        Log::error('There was an error fetching transactions', [
            'message' => $e->getMessage(),
            'trace' => $e->getTrace(),
        ]);

        return response()->json(['error' => 'Unable to fetch transactions'], 500);
    }
}
```

Then this would be the resource:
```php
/**
 * @mixin Transaction
 */
class TransactionResource extends JsonResource
{
    public function toArray($request)
    {
        return [
            'id' => $this->id,
            /** @var array{'amount': float, 'minorAmount': int, 'currency': string} $amount */
            'amount' => money_detail($this->amount),
            'occurredAt' => $this->occurred_at,
            'description' => $this->description,
            'currencyCode' => $this->currency_code,
            'sourceSystem' => $this->source_system,
            'sourceSystemId' => $this->source_system_id,
            'sourceSystemBatchId' => $this->source_system_batch_id,
            'usage' => new UsageResource($this->whenLoaded('usage')),
            'stages' => TransactionStageResource::collection($this->whenLoaded('stages')),
            'images' => TransactionImageResource::collection($this->whenLoaded('images')),
            'notes' => NoteResource::collection($this->whenLoaded('notes')),
            'currentStage' => new TransactionStageResource($this->whenLoaded('currentStage')),
            'identifiers' => TransactionIdentifierResource::collection($this->whenLoaded('identifiers')),
            /** @var array{'amount': float, 'minorAmount': int, 'currency': string} $amount */
            'totalCharged' => $this->when($this->relationLoaded('ledgerEntries'), function () {
                return money_detail($this->totalCharged);
            }),
            /** @var object $metadata Will contain anything included with the metadata when the transaction was originally sent */
            'metadata' => $this->metadata,
            'createdAt' => $this->created_at,
            'updatedAt' => $this->updated_at,
            'accountId' => $this->when($this->whenloaded('account'), $this->account?->id),
        ];
    }
}
```

So instead of the docs page showing all of the relations that are conditional based on `whenLoaded`, it would be great if they only showed the ones that were loaded in the controller via the `with` or `load` methods.

**Comment by Smurrlawa** [comment]
Hello there,
I believe I have an similar issue here.

I have 2 resources that are linked with each other.

1. WebsiteResource
2. WebsiteBackupResource

WebsiteResource has "backups" as an collection of WebsiteBackupResource
WebsiteBackupResource has an link "website" to WebsiteResource

In my WebsiteBackupController@show I want to also include the WebsiteResource for easier handling on FE side.

In my WebsiteController@show I want to display the backups but not the backups to be interlinked again back to Website.
For this purporse I used the whenLoaded inside the WebsiteResource.

`'website' => new WebsiteResource($this->whenLoaded('website')),`

Is there a way to tell Scramble to handle the whenLoaded correctly based on the controller and the input data if it should be supplied or not? Also a DocBloc for this purpose would work.
Even if this doesn work, the loop itself is kinda annoying. Is there a max-depth for this maybe?


This is the display in the docs.
<img width="690" alt="image" src="https://github.com/user-attachments/assets/a7d90740-edb7-47c0-ae77-e5ffebea096d" />

**Reply by romalytvynenko** [reply]
@Smurrlawa 

> Is there a way to tell Scramble to handle the whenLoaded correctly based on the controller and the input data if it should be supplied or not? Also a DocBloc for this purpose would work.

This decision to document things that are conditionally loaded as optional, was made intentionally. If Scramble respects `load` in a controller and generates data structure with it in mind, we cannot make `WebsiteResource` reusable OpenAPI component (as the schema will be different based on what you pass to the constructor).

If you have any ideas on how to maybe explicitly tell Scramble that it should be the separate resource, API ideas for this is welcome.

> Even if this doesn work, the loop itself is kinda annoying. Is there a max-depth for this maybe?

Scramble uses Stoplight Elements for OpenAPI rendering, so there is no max-depth we can control. 

I'm working on a custom UI for OpenAPI rendering and I plan to handle this for sure.

**Reply by Smurrlawa** [reply]
I understand but don't have an idea for this right now...
So I need to use a second Resource if I want to explicitly tell Scramble for this 1 endpoint to remove the "website" in docs?

**Reply by romalytvynenko** [reply]
That would be the simplest solution for now

**Comment by vedmant** [comment]
Any way to tell scramble which relations to show and which to not? Maybe some annotation in the controller? It's pretty important to show in the documentation what data is actually returned, otherwise lots of optional fields shown in the endpoint which tell no useful info.

**Reply by romalytvynenko** [reply]
Not currently, but I'm open to any suggestions regarding the API (how you'd like to tell Scramble how to handle this).

The reason why Scramble is not handling this for now is that Scramble documents JSON resources as reusable schemas and uses them in documentation as references. So while your relations won't show up in ALL endpoints, the generated documentation is still correct by showing them as optional. If I start taking loaded relations into consideration, the JSON resource becomes not reusable across all places.

One thing that comes to my mind is to "inline" the resource schema and simply document is as an object and give users the ability to opt-in into such behavior explicitly. But I don't know, I never had this pain and this need to be honest.

Hope this explanation makes any sense. Let me know what you think! Detailed API suggestions are extremely welcome!

**Reply by vedmant** [reply]
@romalytvynenko Yeah for my use case it will make more sense to just inline the resource schema, maybe something like:

```php
#[InlineResource(['posts', 'posts.comments', 'post.comments.likes', 'posts.comments.likes.users', 'posts.comments.likes.users.media', 'users', 'users.profile', 'users.profile.media'])]
public function show(Event $event)
{
    return new EventResource(
        $event->load(['posts', 'posts.comments', 'post.comments.likes', 'posts.comments.likes.users', 'posts.comments.likes.users.media', 'users', 'users.profile', 'users.profile.media'])
    );
}
```

It might be nice if it can detect eager loading without specifying in the #[InlineResource], maybe also still keeping resources globally for additional reference.

Otherwise with this deep relations nesting current documentation is not much of a use since one endpoint can return only few top relations, other goes way deeper like example above.
